pkg/local_object_storage/metabase/metadata.go

```diff
@@ -0,0 +1,742 @@
+		}
+		if attrs[i] == object.FilterContainerID || attrs[i] == object.FilterID {
+			return nil, "", fmt.Errorf("prohibited attribute %s", attrs[i])
+		}
+	}
+
+	var res []client.SearchResultItem
+	var resCursor []byte
+	var err error
+	if len(fs) == 0 {
+		res, resCursor, err = db.searchUnfiltered(cnr, []byte(cursor), count)
+	} else {
+		res, resCursor, err = db.search(cnr, fs, attrs, []byte(cursor), count)
+	}
+	if err != nil {
+		return nil, "", err
+	}
+	return res, base64.StdEncoding.EncodeToString(resCursor), nil
+}
+
+func (db *DB) search(cnr cid.ID, fs object.SearchFilters, attrs []string, cursor []byte, count uint16) ([]client.SearchResultItem, []byte, error) {
+	seekKey := make([]byte, 1+base64.StdEncoding.DecodedLen(len(cursor))) // TODO: limit?
+	n, err := base64.StdEncoding.Decode(seekKey[1:], cursor)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decode cursor from Base64: %w", err)
+	}
+	seekKey = seekKey[:1+n]
+
+	var res []client.SearchResultItem
+	err = db.boltDB.View(func(tx *bbolt.Tx) error {
+		mb := tx.Bucket(metaBucketKey(cnr))
+		if mb == nil {
+			return nil
+		}
+		var err error
+		res, cursor, err = db.searchInBucket(mb, fs, attrs, seekKey, count)
+		return err
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("view BoltDB: %w", err)
+	}
+	return res, cursor, nil
+}
+
+func (db *DB) searchInBucket(metaBkt *bbolt.Bucket, fs object.SearchFilters, attrs []string,
+	primSeekKey []byte, count uint16) ([]client.SearchResultItem, []byte, error) {
+	// TODO: make as much as possible outside the Bolt tx
+	primMatcher := fs[0].Operation()
+	intPrimMatcher := isNumericOp(primMatcher)
+	notPresentPrimMatcher := primMatcher == object.MatchNotPresent
+	if notPresentPrimMatcher {
+		primSeekKey[0] = metaPrefixID
+	} else if intPrimMatcher {
+		primSeekKey[0] = metaPrefixAttrIDInt
+	} else {
+		primSeekKey[0] = metaPrefixAttrIDPlain
+	}
+
+	primAttr := fs[0].Header()       // attribute emptiness already prevented
+	repeated := len(primSeekKey) > 1 // 1st is prefix
+	var primSeekPrefix []byte
+	var prevResOID, prevResPrimVal []byte
+	if notPresentPrimMatcher {
+		primSeekPrefix = primSeekKey[:1]
+	} else if repeated {
+		ind := bytes.Index(primSeekKey[1:], utf8Delimiter) // 1st is prefix
+		if ind < 0 {
+			return nil, nil, fmt.Errorf("%w: missing delimiter", errInvalidCursor)
+		}
+		if !bytes.Equal(primSeekKey[1:1+ind], []byte(primAttr)) {
+			return nil, nil, fmt.Errorf("%w: wrong primary attribute", errInvalidCursor)
+		}
+		primSeekPrefix = primSeekKey[:1+len(primAttr)+len(utf8Delimiter)]
+		valID := primSeekKey[len(primSeekPrefix):]
+		if len(valID) <= oid.Size {
+			return nil, nil, fmt.Errorf("%w: too small VAL_OID len %d", errInvalidCursor, len(valID))
+		}
+		prevResPrimVal, prevResOID = valID[:len(valID)-oid.Size], valID[len(valID)-oid.Size:]
+	} else {
+		primSeekKey = slices.Concat(primSeekKey[:1], []byte(primAttr), utf8Delimiter)
+		primSeekPrefix = primSeekKey
```

Comment:
What if it's `val < 100`? Won't it be looking at pref_val_100 and next (bigger) ones?

Reply:
note that seek key has no val, so here it'll peek the lowest `pref_attr_` value

but ur question got me thinking: if matcher is, in opposite, `>` or `>=`, why dont we peek `prefix_attr_val` right away? All less values will be skipped anyway, no need to filter them out manually. Worth todo

Reply:
done